src/tango/ip/fd_netlink.c

```diff
@@ -0,0 +1,760 @@
+    uint clazz   = entry->dst_netmask_sz;
+    uint nh_net  = entry->dst_ip_addr;
+    uint dst_net = ip_addr & netmask;
+
+    if( nh_net == dst_net && ( best_idx == -1 || clazz > best_class ) ) {
+      best_idx   = j;
+      best_class = clazz;
+    }
+  }
+
+  if( best_idx < 0L ) {
+    return NULL;
+  } else {
+    return route_table + best_idx;
+  }
+}
+
+
+long
+fd_nl_load_arp_table( fd_nl_t *           nl,
+                      fd_nl_arp_entry_t * arp_table,
+                      ulong               arp_table_cap ) {
+  int fd = nl->fd;
+  if( fd < 0 ) return -1;
+
+  /* format the request */
+
+  /* Request struct */
+  struct {
+    struct nlmsghdr nlh;  /* Netlink header */
+    struct ndmsg    ndm;  /* Payload - neighbor message */
+  } nl_request;
+
+  fd_memset( &nl_request, 0, sizeof( nl_request ) );
+
+  uint seq = nl->seq++;
+
+  nl_request.nlh.nlmsg_type  = RTM_GETNEIGH;  /* We wish to get neighbors */
+  nl_request.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+  nl_request.nlh.nlmsg_len   = NLMSG_LENGTH(sizeof(struct rtmsg));
+  nl_request.nlh.nlmsg_pid   = 0;
+  nl_request.nlh.nlmsg_seq   = seq;
+
+  /* request IPv4 entries */
+  nl_request.ndm.ndm_family  = AF_INET;
+  //nl_request.ndm.ndm_state   = NUD_REACHABLE;
+
+  long sent = send( fd, &nl_request, sizeof(nl_request), 0 );
+  if( sent == -1 ) {
+    FD_LOG_WARNING(( "Unable to make netlink request. Error: %d %s", errno, strerror( errno ) ));
+    return -1;
+  }
+
+  /* receiving */
+  long len;
+
+  /* set alignment such that the returned data is aligned */
+  uchar __attribute__(( aligned(16) )) ibuf[4096]   = {0};
+
+  /* Pointer to the messages head */
+  struct nlmsghdr *h = (struct nlmsghdr *)ibuf;
+
+  /* Try to read the message in case of busy or interrupted call */
+  while(1) {
+    //uchar *     p     = ibuf;
+    //uchar const p_end = ibuf + sizeof(ibuf);
+    do {
+      len = recv( fd, ibuf, sizeof(ibuf), 0 );
+    } while( len < 0 && (errno == EINTR || errno == EAGAIN) );
+
+    if( len < 0 ) {
+      FD_LOG_WARNING(( "netlink recv failed with %d %s", errno, strerror( errno ) ));
+      return -1;
+    }
+
+    if( h->nlmsg_seq == seq ) break;
+  }
+
+  /* index into arp_table */
+  long arp_entry_idx = 0L;
+
+  /* interpret the response */
+
+  long msglen = len;
+
+  /* Iterate through all messages in buffer */
+
+  while( NLMSG_OK(h, msglen) ) {
+    /* are we still within the bounds of the table? */
+    if( arp_entry_idx >= (long)arp_table_cap ) {
+      /* we filled the table */
+      return -1; /* return failure */
+    }
+
+    /* current arp entry */
+    fd_nl_arp_entry_t * entry = arp_table + arp_entry_idx;
+
+    /* clear current entry */
+    fd_memset( entry, 0, sizeof( *entry ) );
+
+    if( h->nlmsg_flags & NLM_F_DUMP_INTR ) {
+      /* function was interrupted - don't switch to new routing table */
+
+      return -1; /* return failure */
+    }
+
+    /* TODO can you spuriously get data from another process?? */
```

Comment:
This is misnamed, but is basically just an ID the kernel echoes back to the process. You can delete the commented code.